logstash-core/src/main/java/org/logstash/secret/store/SecretStoreFactory.java

```diff
@@ -0,0 +1,73 @@
+ * </p>
+ */
+public class SecretStoreFactory {
+
+    public static final String KEYSTORE_ACCESS_KEY = "keystore.pass";
+    //secret stores should create a secret with this as the key and value to identify a logstash secret
+    public final static SecretIdentifier LOGSTASH_MARKER = new SecretIdentifier("logstash-secret-store");
+
+    public final static String ENVIRONMENT_PASS_KEY = "LOGSTASH_SECRET_STORE_PASS";
+
+    /**
+     * Private constructor
+     */
+    private SecretStoreFactory() {
+    }
+
+    private static final Logger LOGGER = LogManager.getLogger(SecretStoreFactory.class);
+
+    /**
+     * Creates a {@link SecretStore} based on the provided configuration
+     *
+     * @param secureConfig The configuration to pass to the implementation
+     * @return
+     */
+    @SuppressWarnings({"unchecked", "JavaReflectionMemberAccess"})
+    static public SecretStore loadSecretStore(SecureConfig secureConfig) {
+        //cheap SPI, if we ever support more then one implementation we should expose it as a setting and push the class name here via the secureConfig
+        String className = System.getProperty("org.logstash.secret.store.SecretStore", "org.logstash.secret.store.backend.JavaKeyStore");
+        try {
+            LOGGER.debug("Attempting to loadSecretStore secret store with implementation: {}", className);
+            Class<? extends SecretStore> implementation = (Class<? extends SecretStore>) Class.forName(className);
+            Constructor<? extends SecretStore> constructor = implementation.getConstructor(SecureConfig.class);
+            addSecretStoreAccess(secureConfig);
+            return constructor.newInstance(secureConfig);
+        } catch (InstantiationException | IllegalAccessException | ClassNotFoundException | NoSuchMethodException e) {
+            throw new SecretStoreException.ImplementationNotFoundException(
+                    String.format("Could not loadSecretStore class %s, please ensure it is on the Java classpath, implements org.logstash.secret.store.SecretStore, and has a 1 " +
+                            "argument constructor that accepts a org.logstash.secret.store.SecureConfig", className), e);
+        } catch (InvocationTargetException ite) {
+            //this is thrown if an exception is thrown from the constructor
+            throw new SecretStoreException.LoadException("Unable to load Logstash secret store", ite);
+        }
+    }
+
+    /**
+     * <p>Adds the credential to the {@link SecureConfig} that is needed to access the {@link SecretStore}. Value read from environment variable "LOGSTASH_KEYSTORE_PASS"</p>
+     *
+     * @param secureConfig The configuration to add the secret store access
+     */
+    private static void addSecretStoreAccess(SecureConfig secureConfig) {
+        String environment = System.getenv(ENVIRONMENT_PASS_KEY);
+
+        char[] pass = null;
+        if (environment != null) {
```

Comment:
This requires the environment variable, but the method still succeeds  when it is not present. We should give some feedback to the user here when a required parameter (this environment variable) is either missing or incorrect.

Reply:
I have added a warning via the CLI w/r/t to this missing value. It will be in a subsequent review.

Reply:
I'm going to forget to check for this, so I can't really evaluate whether or not this is complete :|

Reply:
here is the warning in the CLI : https://github.com/jakelandis/logstash/blob/8353-working/logstash-core/src/main/java/org/logstash/secret/cli/SecretStoreCli.java#L126